### PR TITLE
Adds `CryptoApproveAllowance` preTransactionHandler

### LINKED ISSUE #4226: Implement the `CryptoApproveAllowance` PreTransactionHandler
We need to implement PreTransactionHandler logic that supports `CryptoApproveAllowance` for signing requirements

Implement the below method in `CryptoPreTransactionHandlerImpl`
```
    @Override
    /** {@inheritDoc} */
    public TransactionMetadata preHandleApproveAllowances(TransactionBody txn) {
        throw new NotImplementedException();
    }

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Signed-off-by: Neeharika-Sompalli <neeharika.sompalli@swirldslabs.com>

Fixes #4226 
- Add `CryptoApproveAllowance` preTransactionHandler implementation
- Moved `addIfNotPayer` kind of methods  to `SigTransactionMetadata`, as that seems to reduce duplications.

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java [hunk 1]
```diff
@@ -84,8 +78,94 @@ public ResponseCodeEnum status() {
...
+
+    /**
+     * Checks if the account given is same as payer or if the metadata is already failed. In either
+     * case, no need to lookup tha account's key.
+     *
+     * @param id given account
+     * @return true if the lookup is not needed, false otherwise
+     */
+    private boolean isNotNeeded(final AccountID id) {
+        return id.equals(payer) || id.equals(AccountID.getDefaultInstance()) || status != OK;
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java:153]: This `id.equals(AccountID.getDefaultInstance())` exclusion is specific to `AccountID`'s set in the `CryptoAllowance#owner`, `NftAllowance#owner`, and `TokenAllowance#owner` fields, right?

So on one hand, it seems like we should explicitly limit this exemption to those three contexts.

But on the other hand, if someone leaves a required `AccountID` unset, how could the transaction succeed at consensus? 😁 We're only going to touch the payer account for node + network fees anyways, why bother checking any other signatures?

The same argument would then apply to the `0.0.0` sentinel id, and we could make this:
```
return id.equals(payer)
    || id.equals(AccountID.getDefaultInstance())
    || designatesAccountRemoval(id)
    || status != OK;

...

    private boolean designatesAccountRemoval(AccountID id) {
        return id.getShardNum() == 0
                && id.getRealmNum() == 0
                && id.getAccountNum() == 0
                && id.getAlias().isEmpty();
    }
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java [hunk 2]
```diff
@@ -84,8 +78,94 @@ public ResponseCodeEnum status() {
...
+    }
+
+    /**
+     * Based on the key lookup result, adds the kleys to required keys list if the result is
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java:157]:
```suggestion
     * Given a successful key lookup, adds its key to the required signers. Given a failed key lookup, sets this 
     * SigTransactionMetadata's status to either the failure reason of the lookup; or (if it is non-null), the 
     * requested failureStatus parameter.
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java [hunk 3]
```diff
@@ -84,8 +78,94 @@ public ResponseCodeEnum status() {
...
+     * successful. Else adds failureReason for the status.
+     *
+     * @param result key lookup result
+     * @param failureStatus failure reason for the lookup
+     */
+    private void failOrAddToKeys(
+            final KeyOrLookupFailureReason result, final ResponseCodeEnum failureStatus) {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java:164]: Could mark `@Nullable failureStatus`.

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java:153]: > This id.equals(AccountID.getDefaultInstance()) exclusion is specific to AccountID's set in the CryptoAllowance#owner, NftAllowance#owner, and TokenAllowance#owner fields, right?

yeah , I am not sure, why this condition is added only for allowances ? 🤔 

> The same argument would then apply to the 0.0.0 sentinel id, and we could make this:

Agreed

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/SigTransactionMetadata.java:157]: Fixed. Thanks !

**Review by rbair23 — APPROVED** [review]

**Pull request merged**